### PR TITLE
Field validation e2e tests and GA graduation

### PR DESCRIPTION
**Pull request by kevindelgado**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

* Adds e2e tests for Server Side Unknown Field Validation
* Graduate Server Side Unknown Field Validation to GA
* Drop Enabled() checks for ServerSideFieldValidation feature gate

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

KEP: https://github.com/kubernetes/enhancements/issues/2885

### CODE REVIEW

**Comment by apelisse** [comment]
Yeah, looks good to me, thanks. One minor nit on a test name.

File: test/e2e/apimachinery/field_validation.go [hunk 1]
```diff
@@ -0,0 +1,725 @@
...
+					"containers": [{
+						"name":  "nginx",
+						"image": "nginx:latest"
+					}]
+				}
+			}
+		}
+	}`
+		_, err := client.CoreV1().RESTClient().Post().
+			AbsPath("/apis/apps/v1").
```
Comment (liggitt) [inline test/e2e/apimachinery/field_validation.go:98]: want to make sure this won't trigger more instances of the same issue reported in https://github.com/kubernetes/kubernetes/issues/115276, where running e2e tests with a non `/` server address doesn't work with some tests

I *think* `AbsPath` appends to the server base, but it would be good to verify that

File: pkg/features/kube_features.go [hunk 2]
```diff
@@ -1080,7 +1080,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	genericfeatures.ServerSideApply: {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.29
 
-	genericfeatures.ServerSideFieldValidation: {Default: true, PreRelease: featuregate.Beta},
+	genericfeatures.ServerSideFieldValidation: {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.29
```
Comment (liggitt) [inline pkg/features/kube_features.go:1083]: when locking this to true, drop Enabled(...) checks for this and any code that depends on the feature gate being false

**Review by liggitt — COMMENTED** [review]
just a couple comments, will defer to @apelisse on the test review

**Comment by liggitt** [comment]
verify failure also looks related, though that will drop out when these are no longer conformance tests in this PR

**Comment by apelisse** [comment]
Nice job cleaning-up the docs, thanks!

File: test/e2e/apimachinery/field_validation.go (hunk 1 above)
Comment (kevindelgado) [reply test/e2e/apimachinery/field_validation.go:98]: If I'm understanding that bug correctly, the issue has to do with the `RequestURI` method which is stripping out parts of the path and then adding back the uri blindly.

And you're asking to confirm that `AbsPath()` appends to the request's RESTClient's base directly? (Which is what it seems like it does here, unless I'm misunderstanding the ask: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/request.go#L324)

File: pkg/features/kube_features.go (hunk 2 above)
Comment (kevindelgado) [reply pkg/features/kube_features.go:1083]: done

File: test/e2e/apimachinery/field_validation.go (hunk 1 above)
Comment (liggitt) [reply test/e2e/apimachinery/field_validation.go:98]: yeah, thanks for checking

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go [hunk 3]
```diff
@@ -360,9 +354,9 @@ func (b *builder) buildRoute(root, path, httpMethod, actionVerb, operationVerb s
 		endpoints.AddObjectParams(b.ws, route, &metav1.ListOptions{})
 	case "put", "patch":
 		// TODO: PatchOption added in feature branch but not in master yet
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:356]: totally unrelated to this PR, and should be done in a different PR, but I just noticed this TODO... we're using PatchOptions now, so I think the put and patch cases should split and the patch case should use PatchOptions now

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 4]
```diff
@@ -542,19 +542,19 @@ type CreateOptions struct {
...
+	// prior to v1.23 or prior to v1.27 when the `ServerSideFieldValidation`
+	// feature gate is disabled.
 	// - Warn: This will send a warning via the standard warning response
 	// header for each unknown field that is dropped from the object, and
 	// for each duplicate field that is encountered. The request will
 	// still succeed if there are no other errors, and will only persist
 	// the last of any duplicate fields. This is the default when the
-	// `ServerSideFieldValidation` feature gate is enabled.
+	// `ServerSideFieldValidation` feature gate is enabled (v1.23+) or
+	// locked to enabled in GA (v1.27+)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:557]: I would also simplify this:

```suggestion
	// the last of any duplicate fields. This is the default in v1.23+.
```

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:557]: same comment applies below

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 5]
```diff
@@ -542,19 +542,19 @@ type CreateOptions struct {
...
-	// or duplicate fields, provided that the `ServerSideFieldValidation`
-	// feature gate is also enabled. Valid values are:
+	// or duplicate fields. Valid values are:
 	// - Ignore: This will ignore any unknown fields that are silently
 	// dropped from the object, and will ignore all but the last duplicate
 	// field that the decoder encounters. This is the default behavior
-	// prior to v1.23 and is the default behavior when the
-	// `ServerSideFieldValidation` feature gate is disabled.
+	// prior to v1.23 or prior to v1.27 when the `ServerSideFieldValidation`
+	// feature gate is disabled.
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:550]: same comment applies below

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 6]
```diff
@@ -144,7 +142,7 @@ var (
 				4, 5, 6, 8, 10, 15, 20, 30, 45, 60},
 			StabilityLevel: compbasemetrics.ALPHA,
 		},
-		[]string{"field_validation", "enabled"},
+		[]string{"field_validation"},
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:145]: update the description, since this no longer includes enabled status

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:145]: would be curious to hear from @logicalhan what the preferred approach for something like this is in the future, where we expect `enabled` to eventually be unconditionally true ... should that have been a separate metric, or omitted, or set to `true` unconditionally now?

Comment (kevindelgado) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:145]: updated

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 4 above)
Comment (kevindelgado) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: done

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevindelgado) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: done

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go (hunk 3 above)
Comment (kevindelgado) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:356]: opened https://github.com/kubernetes/kubernetes/pull/115339 stacked on top of this one to avoid conflicts from removing `disabledParams`
